reaper/reaper.go

```diff
@@ -17,141 +17,156 @@
 package reaper
 
 import (
+	"errors"
 	"fmt"
-	"io/ioutil"
+	"github.com/pivotal-cf/service-instance-reaper/cloudfoundry"
+	"io"
 	"time"
-	"github.com/pivotal-cf/service-instance-reaper/httpclient"
-	"net/http"
-	"encoding/json"
 )
 
-// FIXME: support multi-page responses
-const MAXIMUM_RESULTS_PER_PAGE = 100
+const MaximumResultsPerPage = 50
 
-func Reap(authClient httpclient.AuthenticatedClient, apiUrl string, serviceName string, accessToken string, expiryInterval time.Duration, reap bool, fatal func(string)) {
-	bodyReader, statusCode, err := authClient.DoAuthenticatedGet(apiUrl+fmt.Sprintf("/v2/services?q=label:%s", serviceName), accessToken)
-	if err != nil {
-		fatalError("GET /v2/services failed", err, fatal)
-	}
-	if statusCode != http.StatusOK {
-		fatal(fmt.Sprintf("GET /v2/services failed: %d", statusCode))
-	}
+type Reaper struct {
+	cf             cloudfoundry.Client
+	expiryInterval time.Duration
+	reap           bool
+	recursive      bool
+	currentTime    func() time.Time
+	output         io.Writer
+	errorChannel   chan error
+}
 
-	if bodyReader == nil {
-		fatal("GET /v2/services response body missing")
-	}
-	body, err := ioutil.ReadAll(bodyReader)
-	if err != nil {
-		fatalError("Cannot read GET /v2/services response body: %s", err, fatal)
+func NewReaper(cf cloudfoundry.Client, currentTime func() time.Time, output io.Writer) Reaper {
+	return Reaper{
+		cf:          cf,
+		currentTime: currentTime,
+		output:      output,
 	}
+}
 
-	var listServicesResp ListServicesResp
-	err = json.Unmarshal(body, &listServicesResp)
-	if err != nil {
-		fatalError("Invalid GET /v2/services response JSON", err, fatal)
-	}
+func (r Reaper) Reap(serviceName string, expiryInterval time.Duration, reap, recursive bool) error {
```

Comment:
@ojhughes liked your suggestion for changing the signature. We only call `Reap` once and then exit, but I've changed this around from the point of view of what you'd probably want to vary if you were going to make multiple calls. Since `Reaper` holds state, `Reap` takes a non-pointer receiver so it's operating on a copy, to avoid any concurrency issues if things got really weird and we started calling `Reap` from different goroutines.